memstore/recovery.go

```diff
@@ -43,32 +43,28 @@ func (shard *TableShard) ReplayRedoLogs() {
 	utils.GetLogger().With("table", shard.Schema.Schema.Name, "shard", shard.ShardID, "redoLogFile",
 		redoLogFilePersisted, "offset", offsetPersisted).Info("Checkpointed redolog file")
 
-	// Replay redo logs to create LiveStore.
-	nextUpsertBatch := shard.LiveStore.RedoLogManager.NextUpsertBatch()
-
-	for {
-		upsertBatch, redoLogFile, offset := nextUpsertBatch()
-
-		if upsertBatch == nil {
-			break
-		}
-
-		shard.LiveStore.WriterLock.Lock()
-
-		// Put a 0 in maxEventTimePerFile in case this is redolog is full of backfill batches.
-		shard.LiveStore.RedoLogManager.UpdateMaxEventTime(0, redoLogFile)
-
-		// check if this batch has already been backfilled and persisted
-		skipBackfillRows := redoLogFile < redoLogFilePersisted ||
-			(redoLogFile == redoLogFilePersisted && offset <= offsetPersisted)
-		_, err := shard.ApplyUpsertBatch(upsertBatch, redoLogFile, offset, skipBackfillRows)
-
-		shard.LiveStore.WriterLock.Unlock()
-
+	go func() {
+		nextUpsertBatchFunc, err := shard.LiveStore.RedoLogManager.Iterator()
 		if err != nil {
-			utils.GetLogger().With("err", err).Panic("Failed to apply upsert batch during recovery")
+			panic("Fail to start redolog manager")
 		}
-	}
+		for {
+			batchInfo := nextUpsertBatchFunc()
+			if batchInfo == nil {
+				utils.GetLogger().With("table", shard.Schema.Schema.Name, "shard", shard.ShardID).Info("Redolog manager stopped")
```

Comment:
this is error case

Reply:
this is the end of loop, error should be handled in the redolog manager